src/main/java/com/intellij/rt/debugger/agent/CaptureStorage.java

```diff
@@ -383,9 +424,59 @@ public int getRecursionDepth() {
   }
 
   // to be run from the debugger
+
+  /**
+   * Returns the captured stack trace of the current thread.
+   */
   @SuppressWarnings("unused")
   public static String getCurrentCapturedStack(int limit) throws IOException {
-    return wrapInString(CURRENT_STACKS.get().peekLast(), limit);
+    return wrapInString(getStacksForCurrentThread().peekLast(), limit);
+  }
+
+  /**
+   * If storing stack traces for all threads is enabled (`debugger.async.stack.trace.for.all.threads` is true),
+   * returns the captured stack trace of the given thread or null if no stack trace was captured.
+   *
+   * If `debugger.async.stack.trace.for.all.threads` is false,
+   * it only returns the captured stack trace for the current thread and null if no stack trace was captured or if the given thread is not the current thread.
+   */
+  @SuppressWarnings("unused")
+  public static String getCapturedStackForThread(int limit, Thread thread) throws IOException {
+    Deque<CapturedStack> capturedStacks = storeAsyncStackTracesForAllThreads
+            ? THREAD_TO_STACKS_MAP.get(thread)
+            : (thread == Thread.currentThread() ? CURRENT_STACKS.get() : null);
+    if (capturedStacks == null) return null;
+    return wrapInString(capturedStacks.peekLast(), limit);
+  }
+
+  /**
+   * If storing stack traces for all threads is enabled (`debugger.async.stack.trace.for.all.threads` is true),
+   * returns an array of alive threads with their captured stack traces, where array[i][0] is a thread, and array[i][1] is it's stack trace as a string.
+   * A thread is inclued in the array, only if some stack trace was captured for it.
+   *
+   * If `debugger.async.stack.trace.for.all.threads` is false,
+   * only returns the current thread and it's captured stack trace or null if no stack trace was captured.
+   */
+  public static Object[][] getAllCapturedStacks(int limit) throws IOException {
```

Comment:
It is still to many requests, better have one array with odd-even being thread and stack

Reply:
we still have to request string value for every string separately :( maybe it is better to send an array of threads and a combined string of stacktraces